AI content updates (#17994)
ethereum/ethereum-org-website: public/content/ai-agents/index.md

summaryPoint2: Verify agent behavior onchain using zero-knowledge machine learning (zkML) and trusted execution environments (TEEs)
summaryPoint3: Enable machine-to-machine payments and agent commerce with stablecoin rails and open payment protocols
buttons:
  - content: Getting started
    href: /ai-agents/getting-started/
  - content: Browse the hub
    toId: ai-agent-builder-hub-sections
    isSecondary: true
---

AI agents are software programs that observe their environment, make decisions, and take actions, including sending transactions, paying for services, and interacting with smart contracts, without requiring human input on every step.

Ethereum provides the infrastructure these agents need: programmable wallets that enforce spending limits, verifiable computation that proves an agent followed its rules, open payment rails for machine-to-machine commerce, and a permissionless identity layer that any application can verify. These are not properties that a centralized server can replicate, because they depend on a network that no single party controls.

This builder hub is the starting point for developers building production AI agents on Ethereum.

> <Alert variant="info">

Reviewer: Add vertical margin so this banner gets visible breathing room above and below (the MDX `<Alert>` sits flush against the surrounding paragraph and the next `<h2>` otherwise).

```suggestion
<Alert variant="info" className="my-8">
```

## What the AI agent builder hub covers {#what-the-ai-agent-hub-covers}

Ethereum provides a unique set of technical foundations for deploying production AI agents. 

The AI agent builder hub covers each of these core building blocks in depth:

- **Programmable wallets**: Agents can hold and transfer digital assets directly. A [smart account](/account-abstraction/) functions as the agent's onchain identity and treasury, with spending limits, allowlists, and session keys enforced at the contract level. See [Agent wallets](/ai-agents/wallets/).
- **Verifiable inference**: An agent running offchain is a black box. By anchoring key decisions to onchain proofs via zkML or TEEs, developers can give counterparties a way to confirm that the agent followed its stated logic. See [Verification](/ai-agents/verification/).
- **Machine payments**: Autonomous agents need to pay for compute and data without human intervention. Stablecoins and open protocols like x402 provide a programmable settlement layer for agent commerce. See [Payments](/ai-agents/payments/).

> - **Agent identity**: When agents interact with humans or other agents, counterparties need a way to verify what an agent is and what its track record is using onchain registries like ERC-8004. See [Identity](/ai-agents/identity/).

Reviewer: Convert the four-item bullet list into a `CardGrid` to match the pattern used on other use-case pages (`real-world-assets`, `payments`, `prediction-markets`). Gives each capability visual parity and stronger affordance for the per-topic deep link.

Note: drops the inline `/account-abstraction/` link in the wallets description (since the `/ai-agents/wallets/` destination covers smart accounts in depth). If you want to keep it, change that `Card`'s description to a JSX fragment: `description={<>...<a href="/account-abstraction/">smart account</a>...</>}`.

```suggestion
<CardGrid className="grid grid-cols-[repeat(auto-fill,_minmax(min(100%,_280px),_1fr))] gap-8 my-8">
  <Card emoji=":closed_lock_with_key:" title="Programmable wallets" description="Agents can hold and transfer digital assets directly. A smart account functions as the agent's onchain identity and treasury, with spending limits, allowlists, and session keys enforced at the contract level.">
    <a href="/ai-agents/wallets/">See Agent wallets</a>
  </Card>
  <Card emoji=":mag:" title="Verifiable inference" description="An agent running offchain is a black box. By anchoring key decisions to onchain proofs via zkML or TEEs, developers can give counterparties a way to confirm that the agent followed its stated logic.">
    <a href="/ai-agents/verification/">See Verification</a>
  </Card>
  <Card emoji=":money_with_wings:" title="Machine payments" description="Autonomous agents need to pay for compute and data without human intervention. Stablecoins and open protocols like x402 provide a programmable settlement layer for agent commerce.">
    <a href="/ai-agents/payments/">See Payments</a>
  </Card>
  <Card emoji=":bust_in_silhouette:" title="Agent identity" description="When agents interact with humans or other agents, counterparties need a way to verify what an agent is and what its track record is using onchain registries like ERC-8004.">
    <a href="/ai-agents/identity/">See Identity</a>
  </Card>
</CardGrid>
```

---
title: Build AI agents on Ethereum
metaTitle: AI agent builder hub | ethereum.org
description: A developer hub for building AI agents on Ethereum. Covers wallets, verifiable computation, machine payments, agent identity, and the L2 ecosystem.
lang: en
template: ai-agents
emoji: ":robot:"
sidebarDepth: 2
image: /images/ai-agents/hero-image.png
alt: Developer working with AI agent architecture on Ethereum

> summaryPoint1: Deploy autonomous agents with programmable wallets, spending limits, and session keys

Reviewer: Trim to ≤8 words; keep keywords; parallel structure.

```suggestion
summaryPoint1: Programmable wallets with spending limits and session keys
```

> summaryPoint1: Deploy autonomous agents with programmable wallets, spending limits, and session keys

Reviewer: Trim to ≤8 words; keep keywords; parallel structure.

```suggestion
summaryPoint1: Programmable wallets with spending limits and session keys
```

> summaryPoint2: Verify agent behavior onchain using zero-knowledge machine learning (zkML) and trusted execution environments (TEEs)

Reviewer: Trim to ≤8 words; keep keywords; parallel structure.

```suggestion
summaryPoint2: Verifiable behavior via zkML proofs and TEE attestations
```

> summaryPoint2: Verify agent behavior onchain using zero-knowledge machine learning (zkML) and trusted execution environments (TEEs)

Reviewer: Trim to ≤8 words; keep keywords; parallel structure.

```suggestion
summaryPoint2: Verifiable behavior via zkML proofs and TEE attestations
```

> summaryPoint3: Enable machine-to-machine payments and agent commerce with stablecoin rails and open payment protocols

Reviewer: Trim to ≤8 words; keep keywords; parallel structure.

```suggestion
summaryPoint3: Machine-to-machine payments using stablecoins and open protocols
```

> summaryPoint3: Enable machine-to-machine payments and agent commerce with stablecoin rails and open payment protocols

Reviewer: Trim to ≤8 words; keep keywords; parallel structure.

```suggestion
summaryPoint3: Machine-to-machine payments using stablecoins and open protocols
```

    <a href="/ai-agents/wallets/">See Agent wallets</a>
  </Card>
  <Card emoji=":mag:" title="Verifiable inference" description="An agent running offchain is a black box. By anchoring key decisions to onchain proofs via zkML or TEEs, developers can give counterparties a way to confirm that the agent followed its stated logic.">
    <a href="/ai-agents/verification/">See Verification</a>
  </Card>
  <Card emoji=":money_with_wings:" title="Machine payments" description="Autonomous agents need to pay for compute and data without human intervention. Stablecoins and open protocols like x402 provide a programmable settlement layer for agent commerce.">
    <a href="/ai-agents/payments/">See Payments</a>
  </Card>
  <Card emoji=":bust_in_silhouette:" title="Agent identity" description="When agents interact with humans or other agents, counterparties need a way to verify what an agent is and what its track record is using onchain registries like ERC-8004.">
    <a href="/ai-agents/identity/">See Identity</a>
  </Card>

> </CardGrid>

Reviewer: Use the new `HubCard` component (uses the same Card primitives as `/use-cases` and `/learn`) so this hub matches the visual rhythm of the rest of the site. The `<CardGrid>` className override is no longer needed; the layout default drives spacing.

```suggestion
<CardGrid>
  <HubCard
    emoji=":closed_lock_with_key:"
    title="Programmable wallets"
    description="Agents can hold and transfer digital assets directly. A smart account functions as the agent's onchain identity and treasury, with spending limits, allowlists, and session keys enforced at the contract level."
    href="/ai-agents/wallets/"
    ctaLabel="See Agent wallets"
  />
  <HubCard
    emoji=":mag:"
    title="Verifiable inference"
    description="An agent running offchain is a black box. By anchoring key decisions to onchain proofs via zkML or TEEs, developers can give counterparties a way to confirm that the agent followed its stated logic."
    href="/ai-agents/verification/"
    ctaLabel="See Verification"
  />
  <HubCard
    emoji=":money_with_wings:"
    title="Machine payments"
    description="Autonomous agents need to pay for compute and data without human intervention. Stablecoins and open protocols like x402 provide a programmable settlement layer for agent commerce."
    href="/ai-agents/payments/"
    ctaLabel="See Payments"
  />
  <HubCard
    emoji=":bust_in_silhouette:"
    title="Agent identity"
    description="When agents interact with humans or other agents, counterparties need a way to verify what an agent is and what its track record is using onchain registries like ERC-8004."
    href="/ai-agents/identity/"
    ctaLabel="See Identity"
  />
</CardGrid>
```

Pairs with the `HubCard` definition suggestion in `src/layouts/md/AiAgents.tsx`.

---
title: Build AI agents on Ethereum
metaTitle: AI agent builder hub | ethereum.org
description: A developer hub for building AI agents on Ethereum. Covers wallets, verifiable computation, machine payments, agent identity, and the L2 ecosystem.
lang: en
template: ai-agents
emoji: ":robot:"
sidebarDepth: 2
image: /images/ai-agents/hero-image.png
alt: Developer working with AI agent architecture on Ethereum
summaryPoint1: Programmable wallets with spending limits and session keys

> summaryPoint2: Verifiable behavior via zkML proofs and TEE attestations

Reviewer 2: Verifiable behavior via TEE attestations today, with zkML proofs emerging for smaller models.

summaryPoint3: Machine-to-machine payments using stablecoins and open protocols
buttons:
  - content: Getting started
    href: /ai-agents/getting-started/
  - content: Browse the hub
    toId: ai-agent-builder-hub-sections
    isSecondary: true
---

AI agents are software programs that observe their environment, make decisions, and take actions, including sending transactions, paying for services, and interacting with smart contracts, without requiring human input on every step.

> Ethereum provides the infrastructure these agents need: programmable wallets that enforce spending limits, verifiable computation that proves an agent followed its rules, open payment rails for machine-to-machine commerce, and a permissionless identity layer that any application can verify. These are **not properties that a centralized server can replicate**, because they depend on a network that no single party controls.

Reviewer 2: - "verifiable computation that proves an agent followed its rules": Reads like Ethereum provides this already today
- "not properties that a centralized server can replicate": a centralized server can publish enforce spending limits, zk proofs, TEE attestations --> better framing about who controls the rules and can change or shut them down

This builder hub is the starting point for developers building production AI agents on Ethereum.

<Alert variant="info" className="my-8">
<AlertContent>
<p className="mt-0"><strong>New to AI on Ethereum?</strong></p>
<p className="mt-2">If you want to learn how AI and Ethereum work together as a use case, visit the <a href="/decentralized-ai/">Decentralized AI</a> overview page.</p>
</AlertContent>
</Alert>

## AI agents vs AI bots {#ai-agents-vs-bots}

> |                      | AI agents                                                                      | AI bots                                        |

Reviewer 2: Maybe just bots? AI bots contradicts with AI agents

| -------------------- | ------------------------------------------------------------------------------ | ---------------------------------------------- |

> | **Decision-making**  | Autonomous; adapts to new inputs and conditions                                | Rule-based; executes fixed, pre-programmed logic |

Reviewer 2: - "Rule-based; executes fixed, pre-programmed logic": thats more an algorithm not AI, maybe drop AI in front of bot

| **Learning**         | Can update behavior from real-time data and feedback                           | Operates on pre-trained data or hardcoded rules  |

> | **Onchain identity** | Holds a programmable wallet, can sign and send transactions                    | Typically reads data; does not hold assets       |

Reviewer 2: Also bots can hold wallets (e.g. trading bots)

> | **Composability**    | Can discover and interact with any open protocol                                | Limited to the APIs it was explicitly integrated with |

Reviewer 2: dynamic discovery vs. static integration

> | **Guardrails**       | Enforced by smart contract logic (spending limits, session keys, allowlists)   | Enforced by the operator's server configuration  |

Reviewer 2: a bot could also hold a smart contract with sessions keys, etc.

> | **Verification**     | Behavior can be proven onchain via zkML or TEE attestations                    | Behavior is attested only by the operator        |

Reviewer 2: Bots can also be verified

    href="/ai-agents/payments/"
    ctaLabel="See Payments"
  />
  <HubCard
    emoji=":bust_in_silhouette:"
    title="Agent identity"
    description="When agents interact with humans or other agents, counterparties need a way to verify what an agent is and what its track record is using onchain registries like ERC-8004."
    href="/ai-agents/identity/"
    ctaLabel="See Identity"
  />
</CardGrid>

## AI agent infrastructure today {#ecosystem-today}

The standards and protocols that power AI agents on Ethereum have crossed from research into production.

| Standard or protocol | Status |
| :--- | :--- |
| **EIP-7702** — Smart account delegation for existing EOAs | Live since the Pectra upgrade, May 2025 |
| **ERC-4337** — Smart account standard | Production EntryPoint deployments on all major networks (v0.7 widely adopted; v0.8, v0.9 also live) |
| **ERC-8004** — Onchain agent identity registry | Draft — in active adoption; live on 20+ chains since January 2026 |
| **x402** — Machine-to-machine HTTP payments | Production-ready since Q1 2026 |

> | **L2 transaction cost** | $0.001–$0.003 per operation |

Reviewer 2: Maybe: sub-cent to low-cents per operation
